### PR TITLE
[SPARK-12340][SQL]fix Int overflow in the SparkPlan.executeTake, RDD.take and AsyncRDDActions.takeAsync

### PR DESCRIPTION
**Pull request by QiangCai**
I have closed pull request https://github.com/apache/spark/pull/10487. And I create this pull request to resolve the problem.

spark jira
https://issues.apache.org/jira/browse/SPARK-12340

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -2028,4 +2028,18 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
+    val rdd = sqlContext.sparkContext.parallelize(1 to 3 , 3 )
+    
+    rdd.toDF("key").registerTempTable("spark12340")
+    checkAnswer(
+      sql("select key from spark12340 limit 2147483638"),    
+      Row(1) :: Row(2) :: Row(3) :: Nil
+    )
+    
+    assert(rdd.take(2147483638).size === 3)
+
```
Comment (sarutak) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: I don't think a blank line is not needed here.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -2028,4 +2028,18 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
+    
+    rdd.toDF("key").registerTempTable("spark12340")
+    checkAnswer(
+      sql("select key from spark12340 limit 2147483638"),    
+      Row(1) :: Row(2) :: Row(3) :: Nil
+    )
+    
+    assert(rdd.take(2147483638).size === 3)
+    
+    assert(rdd.takeAsync(2147483638).get.size === 3)
```
Comment (sarutak) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Does this test case pass the compilation? `2147483638` is not within the range of `Int`.
Should we say `Int.MaxValue` ?

**Comment by QiangCai** [comment]
I have removed some blank lines.

**Comment by srowen** [comment]
@QiangCai the problem isn't blank lines but whitespace at the end of your lines.

**Comment by QiangCai** [comment]
@srowen I have removed some whitespaces.

**Comment by QiangCai** [comment]
@srowen  I have no idea how to resolve this error. Would you help me?

**Comment by srowen** [comment]
@QiangCai I think the test failures are unrelated. However before we can retest you'll have to rebase as there is a merge conflict now.

**Comment by QiangCai** [comment]
@srowen I have rebased from master and resolved all conflicts.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 3]
```diff
@@ -2028,6 +2028,7 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       Row(false) :: Row(true) :: Nil)
   }
 
+<<<<<<< 1cdc42d2b99edfec01066699a7620cca02b61f0e
```
Comment (srowen) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: This shouldn't be here

Comment (QiangCai) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: I have remove it.

**Comment by QiangCai** [comment]
@srowen I find error message in test bulid log, a OutOfMemoryError exception has happened. The code in 71 line of  the file AsyncRDDActions.scala is  "val results = new ArrayBuffer[T](num) ", because num(2147483638) is too large, JVM can't allocate enough memory space. I think I should modify unit test to smaller num in 2079 line of SQLQuerySuite.scala.

error message:
[info] Exception encountered when attempting to run a suite with class name: org.apache.spark.sql.SQLQuerySuite *** ABORTED *** (31 seconds, 447 milliseconds)
[info]   java.lang.OutOfMemoryError: Java heap space
[info]   at scala.collection.mutable.ResizableArray$class.$init$(ResizableArray.scala:32)
[info]   at scala.collection.mutable.ArrayBuffer.<init>(ArrayBuffer.scala:47)
[info]   at org.apache.spark.rdd.AsyncRDDActions$$anonfun$takeAsync$1.apply(AsyncRDDActions.scala:71)
[info]   at org.apache.spark.rdd.AsyncRDDActions$$anonfun$takeAsync$1.apply(AsyncRDDActions.scala:66)
[info]   at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:150)
[info]   at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:111)
[info]   at org.apache.spark.rdd.RDD.withScope(RDD.scala:316)
[info]   at org.apache.spark.rdd.AsyncRDDActions.takeAsync(AsyncRDDActions.scala:66)
[info]   at org.apache.spark.sql.SQLQuerySuite$$anonfun$132.apply$mcV$sp(SQLQuerySuite.scala:2079)
[info]   at org.apache.spark.sql.SQLQuerySuite$$anonfun$132.apply(SQLQuerySuite.scala:2071)
[info]   at org.apache.spark.sql.SQLQuerySuite$$anonfun$132.apply(SQLQuerySuite.scala:2071)
......

**Comment by sarutak** [comment]
Why the instance of `ArrayBuffer` in `AsyncRDDActions#takeAsync` is created with initial size?
On the other hand, the instance of `ArrayBuffer` in `RDD#take` is created without initial size.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -2067,4 +2067,16 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       )
     }
   }
+
+  test("SPARK-12340: overstep the bounds of Int in SparkPlan.executeTake"){
```
Comment (sarutak) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: If you have a chance to modify again, please insert a white space between `)` and `{`.

**Comment by QiangCai** [comment]
@sarutak  Maybe we have found another bug. I will try to fix it.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 4 above)
Comment (QiangCai) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: I will do it.

**Comment by QiangCai** [comment]
I have remove the initial size of the ArrayBuffer instance. And the default size is 16.

**Comment by QiangCai** [comment]
I think I have resolved this problem.

**Comment by srowen** [comment]
LGTM

**Comment by sarutak** [comment]
Merging this into `master` and `branch-1.6`. Thanks @QiangCai !

**Comment by sarutak** [comment]
@QiangCai We have many conflicts against `branch-1.6` so I'd merge this into only `master` for now.
If you want merge this into `branch-1.6` please feel free open another PR.

**Comment by QiangCai** [comment]
OK. I have created another PR https://github.com/apache/spark/pull/10619.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 5]
```diff
@@ -1291,11 +1291,11 @@ abstract class RDD[T: ClassTag](
     } else {
       val buf = new ArrayBuffer[T]
       val totalParts = this.partitions.length
-      var partsScanned = 0
+      var partsScanned = 0L
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: why is this change necessary? When can partsScanned go above 2B?

Comment (sarutak) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: Ah, you're right. `partScanned` cannot exceed the value of `totalParts`.
I'll return it to `Int`.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: I think there is a legit problem here. Imagine `totalParts` is close to `Int.MaxValue`, and imagine `partsScanned` is close to `totalParts`. Adding `p.size` to it below could cause it to roll over. I think this change is needed.

Comment (rxin) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: That's never possible -- if we have anywhere near 2B partitions, the scheduler won't be fast enough to schedule them. As a matter of fact, if we have anywhere larger than a few millions, the scheduler will likely crash.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: Fair point, in practice this all but certainly won't happen. Note that this patch was already committed to `master` making this a `Long`. It doesn't hurt and is very very theoretically more correct locally. I suppose I don't think it's worth updating again, but I do not feel strongly about it.

Comment (rxin) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: I'd prefer to change it back since it is so little work, in case this starts a trend to change all ints to longs for no reason, and also raise questions about why this can be greater than int.max when we read this code in the future.

Also @srowen even if totalParts is close to int.max, I don't think partsScanned can be greater than int.max because we never scan more parts than the number of parts available.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 6]
```diff
@@ -2067,4 +2067,16 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       )
     }
   }
+
+  test("SPARK-12340: overstep the bounds of Int in SparkPlan.executeTake") {
+    val rdd = sqlContext.sparkContext.parallelize(1 to 3 , 3 )
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: also remove the extra space before comma here.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 7]
```diff
@@ -2067,4 +2067,16 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
+    rdd.toDF("key").registerTempTable("spark12340")
+    checkAnswer(
+      sql("select key from spark12340 limit 2147483638"),
+      Row(1) :: Row(2) :: Row(3) :: Nil
+    )
+    assert(rdd.take(2147483638).size === 3)
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: we should have a unit test in RDDSuite for the rdd tests, not in SQLQuerySuite.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala (hunk 5 above)
Comment (rxin) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: Ah ok you were referring to partsScanned + numPartsToTry - we should just cast that to long to minimize the impact.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 6 above)
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: for the sql part i'd just move this into the existing limit test case, and add a line of comment explaining this.

also you should explain in the comment why 2147483638 is chosen as a value.

**Comment by rxin** [comment]
@QiangCai it would be great if you can submit a new pull request to address the comments. Thanks.
